src/openms_gui/source/VISUAL/SpectrumCanvas.cpp

```diff
@@ -391,60 +391,120 @@ namespace OpenMS
 
   bool SpectrumCanvas::addLayer(ExperimentSharedPtrType map, ODExperimentSharedPtrType od_map, const String & filename)
   {
-    layers_.resize(layers_.size() + 1);
-    layers_.back().param = param_;
-    layers_.back().filename = filename;
-    layers_.back().setPeakData(map);
-    layers_.back().setOnDiscPeakData(od_map);
-
-    if (layers_.back().getPeakData()->getChromatograms().size() != 0 
-        && layers_.back().getPeakData()->size() != 0)
+    LayerData new_layer;
+    new_layer.param = param_;
+    new_layer.filename = filename;
+    new_layer.name = QFileInfo(filename.toQString()).baseName();
+    new_layer.setPeakData(map);
+    new_layer.setOnDiscPeakData(od_map);
+
+    if (new_layer.getPeakData()->getChromatograms().size() != 0 
+        && new_layer.getPeakData()->size() != 0)
     {
       // TODO : handle this case better
       OPENMS_LOG_WARN << "Your input data contains chromatograms and spectra, falling back to display spectra only." << std::endl;
     }
 
-    if (layers_.back().getPeakData()->getChromatograms().size() != 0 
-        && layers_.back().getPeakData()->size() == 0)
+    if (new_layer.getPeakData()->getChromatograms().size() != 0 
+        && new_layer.getPeakData()->size() == 0)
     {
-      layers_.back().type = LayerData::DT_CHROMATOGRAM;
+      new_layer.type = LayerData::DT_CHROMATOGRAM;
     }
     else
     {
-      layers_.back().type = LayerData::DT_PEAK;
+      new_layer.type = LayerData::DT_PEAK;
+    }
+
+    // insert after last peak or chomatogram layer
+    for (auto it = layers_.rbegin(); it != layers_.rend(); ++it)
+    {
+      if (it->type == LayerData::DT_PEAK 
+        || it->type == LayerData::DT_CHROMATOGRAM)
+      {
+        layers_.insert(it.base(), new_layer);
```

Comment:
in order to get the move insert , e.g. https://en.cppreference.com/w/cpp/container/vector/insert (2) you need to do this to at least prepare for this option:  
```suggestion
        layers_.insert(it.base(), std::move(new_layer));
```

Reply:
PS: T must meet the requirements of MoveAssignable and MoveInsertable in order to use overload (2).

Reply:
> PS: T must meet the requirements of MoveAssignable and MoveInsertable in order to use overload (2).

so this means that `LayerData` should have a move assignment operator and either a move constructor or a copy constructor that is `const -ref`. It currently has neither. Another idea would be that the layer would keep track of `shared_ptr` of the data and no data ever gets copied when the layer gets copied

Reply:
The bigger problem here is the `insert()`, which will copy all layers behind the insertion point to their new destination (or copy the whole vector upon resize). All because we lack the move assignment. So this would be really nice to have.